interfaces/builtin/microstack_support.go

```diff
@@ -0,0 +1,274 @@
+/sys/module/kvm_hv/parameters/nested r, # PPC64.
+/sys/module/kvm/parameters/nested r, # S390.
+
+# Used by libvirt (cgroup-related):
+/sys/fs/cgroup/** rw,
+/sys/fs/cgroup/*/machine/qemu-*/** rw,
+/sys/fs/cgroup/unified/cgroup.controllers r,
+/proc/self/cgroup r,
+/proc/cgroups r,
+
+# Used by libvirt.
+/proc/cmdline r,
+/proc/filesystems r,
+/proc/mtrr w,
+/proc/@{pids}/environ r,
+/proc/@{pids}/mountinfo r,
+/proc/@{pids}/mounts r,
+/proc/@{pids}/sched r,
+/proc/@{pids}/stat r,
+
+# Per man(5) proc, the kernel enforces that a thread may
+# only modify its comm value or those in its thread group.
+owner /proc/@{pid}/task/@{tid}/comm rw,
+
+/proc/*/cmdline r,
+/proc/sys/net/** r,
+/proc/version r,
+/proc/*/stat r,
+/proc/*/mounts r,
+/proc/*/status r,
+/proc/*/ns/net r,
+
+
+# Used by libvirt to work with network devices created for VMs.
+# E.g. getting operational state and speed of tap devices.
+/sys/class/net/tap*/* rw,
+
+# for usb access
+/dev/bus/usb/ r,
+/sys/bus/ r,
+/sys/class/ r,
+# For hostdev access. The actual devices will be added dynamically
+/sys/bus/usb/devices/ r,
+/sys/devices/**/usb[0-9]*/** r,
+# libusb needs udev data about usb devices (~equal to content of lsusb -v)
+/run/udev/data/+usb* r,
+/run/udev/data/c16[6,7]* r,
+/run/udev/data/c18[0,8,9]* r,
+
+# Libvirt needs access to the PCI config space in order to be able to reset devices.
+/sys/devices/pci*/**/config rw,
+
+# for vfio hotplug on systems without static vfio (LP: #1775777)
+/dev/vfio/vfio rw,
+/dev/vfio/* rw,
+
+# Allow querying the max number of segments for a block device by Libvirt
+# (see QEMU commit 9103f1ce and Linux/block/queue-sysfs.txt).
+/sys/devices/**/block/*/queue/max_segments r,
+
+# required by libpmem init to fts_open()/fts_read() the symlinks in
+# /sys/bus/nd/devices
+/ r, # harmless on any lsb compliant system
+/sys/bus/nd/devices/{,**/} r,
+
+# For ppc device-tree access by Libvirt.
+/proc/device-tree/ r,
+/proc/device-tree/** r,
+/sys/firmware/devicetree/** r,
+
+# "virsh console" support
+/dev/pts/* rw,
+# Used by libvirt.
+/dev/ptmx rw,
+# spice
+owner /{dev,run}/shm/spice.* rw,
+
+# Used by libvirt to create lock files for /dev/pts/<num> devices
+# when handling virsh console access requests.
+/run/lock/ r,
+/run/lock/snap.@{SNAP_INSTANCE_NAME}/ rw,
+/run/lock/snap.@{SNAP_INSTANCE_NAME}/** mrwklix,
+
+# allow connect with openGraphicsFD to work
+unix (send, receive) type=stream addr=none peer=(label=libvirtd),
+unix (send, receive) type=stream addr=none peer=(label=/snap/microstack/current/usr/sbin/libvirtd),
+
+# Allow running utility processes under the specialized AppArmor profiles.
+# These profiles will prevent utility processes escaping confinement.
+capability mac_admin,
+
+# MicroStack services such as libvirt use has a server/client design where
+# unix sockets are used for IPC.
+capability chown,
+
+# MicroStack will also use privilege separation when running utility processes
+capability setuid,
+capability setgid,
```

Comment:
With these 3 capabilities, are you running the VMs as root? As 'snap_daemon'? (See https://forum.snapcraft.io/t/system-usernames/13386). It would be nice if microstack started them as non-root as 'snap_daemon:snap_daemon'. If you did that, you can remove these since they would be included in the template.

If not, what are you chowning and setuid/setgid'ing to?